### PR TITLE
GPII-3398: Replace express-pouchdb with CouchDB running in a Docker container.

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [GPII-3398](https://issues.gpii.net/browse/GPII-3398) for details.

@cindyli, we should discuss this approach with regards to your work in our meeting on Tuesday.

### CODE REVIEW

**Comment by the-t-in-rtf** [comment]
There are lots of advantages of this approach (ease of updating, faithfulness in comparison with what how we work with CouchDB in production, reduced maintenance, improved security).  However, there is a significant amount of time involved in building up and tearing down the docker container that runs CouchDB.   A full run of `all-tests` took 4.4 seconds with express-pouchdb, with the new Docker container it takes 21.6 seconds.  We should also discuss how best to support instances that are meant to persist between restarts, we might "reuse and reset" a single container for all tests, for example, and use a similar mechanism (without the reset) for Cindy's use case.

File: tests/js/docker-couchdb-tests/basic-tests.js [hunk 1]
```diff
@@ -14,24 +14,38 @@ require("gpii-express");
...
+gpii.tests.pouch.basic.checkRecordAndStartDelete = function (response, body, expectedStatus, expectedBody, deleteRequest) {
+    var record = JSON.parse(body);
+    gpii.test.pouch.checkResponse(response, body, expectedStatus, expectedBody);
+
+    // DELETE requests must reference a specific revision, as in:
+    // DELETE /recipes/FishStew?rev=1-9c65296036141e575d32ba9c034dd3ee
+    deleteRequest.send({}, { termMap: { rev: record._rev } });
+};
 
 fluid.defaults("gpii.tests.pouch.basic.caseHolder", {
```
Comment (amb26) [inline tests/js/docker-couchdb-tests/basic-tests.js:31]: Odd that pouch has been replaced with pouchdb in the base gradeName but not throughout

File: package.json [hunk 2]
```diff
@@ -9,38 +9,32 @@
...
+        "express": "4.16.4",
+        "gpii-express": "1.0.15",
         "graceful-fs": "4.1.11",
-        "infusion": "3.0.0-dev.20180222T160835Z.6e1311a",
-        "lie": "3.2.0",
-        "memdown": "2.0.0",
-        "pouchdb": "6.4.3",
+        "infusion": "3.0.0-dev.20181024T171748Z.1fa91bed9",
+        "lie": "3.3.0",
+        "pouchdb": "7.0.0",
```
Comment (amb26) [inline package.json]: I see that pouchdb is still present as a dependency - is it used in some capacity?

File: src/js/harness.js [hunk 3]
```diff
@@ -1,69 +1,424 @@
...
+
+/**
+ *
+ * Create the required databases and populate them with data.
+ *
+ * @param {Object} that - The component itself.
+ * @return {Promise} - A `fluid.promise` instance that will be resolved when the data is loaded or rejected if an error occurs.
+ *
+ */
+gpii.pouch.harness.loadData = function (that) {
```
Comment (amb26) [inline src/js/harness.js]: This function is too huge to survive as a monolith - please factor appropriately

File: package.json [hunk 4]
```diff
@@ -9,38 +9,32 @@
...
-        "gpii-express": "1.0.13",
+        "express": "4.16.4",
+        "gpii-express": "1.0.15",
         "graceful-fs": "4.1.11",
-        "infusion": "3.0.0-dev.20180222T160835Z.6e1311a",
-        "lie": "3.2.0",
-        "memdown": "2.0.0",
-        "pouchdb": "6.4.3",
+        "infusion": "3.0.0-dev.20181024T171748Z.1fa91bed9",
+        "lie": "3.3.0",
```
Comment (amb26) [inline package.json]: I can't see where this dependency is used

**Comment by amb26** [comment]
This is definitely an approach we want to go with, but I am concerned whether we can use it in all cases. As well as the risk of slowing down test runs against universal, I am wondering whether the "Docker Toolbox" required for earlier versions of windows will integrate adequately with this approach or whether I will then always need to run universal's tests from inside a VM.

Certainly investigating whether we can reuse the same Docker image across resets, together with a suitably trustworth "burn Couch to the ground" primitive seems like a good research avenue.

**Comment by the-t-in-rtf** [comment]
I could use your help verifying the supportability of Windows 7.  Running any individual test in the docker-couchdb-tests folder should be enough to confirm whether it works.

Also, I am already researching the speed issues, it's a bit more complex than just the buildup/teardown.  Here are sample log messages from a single testEnvironment:

```
14:53:05.013:  Starting container.
14:53:05.655:  Loading data.
14:53:08.446:  Harness ready.
...
14:53:08.574:  Removing container.
14:53:09.495:  Verifying that container has been removed.
14:53:09.575:  Removing container.
14:53:09.622:  Checking for existing gpii-pouchdb containers.
```

It's around 1.5s for buildup/teardown (.6s + .9s), and around 2.8s for data loading.  So if we only refresh the container, we only get back roughly a third of the time.  Adding a mechanism to avoid reloading particular data sets would get us back more, but we increase the risk of persisting changes from one test environment to the next.

File: package.json (hunk 2 above)
Comment (the-t-in-rtf) [reply package.json]: Yes, it's used on the browser side in Simon (and possibly Cindy's) work.

File: package.json (hunk 4 above)
Comment (the-t-in-rtf) [reply package.json]: Yup, it's gone.

File: package.json (hunk 2 above)
Comment (the-t-in-rtf) [reply package.json]: In this package it's used [in the browser tests](https://github.com/the-t-in-rtf/gpii-pouchdb/blob/GPII-3398/tests/js/pouchdb-component-tests/html/index.html#L29).

File: tests/js/docker-couchdb-tests/basic-tests.js (hunk 1 above)
Comment (the-t-in-rtf) [reply tests/js/docker-couchdb-tests/basic-tests.js:31]: Seems like the waters are pretty muddy.  Almost all of the code in `./src` uses "pouch", but a lot of the tests, including the stuff that people are already using (like the `testEnvironment`) use "pouchdb".

I would rather not break people's existing tests by renaming, but I'm already increasingly uncomfortable with either the "pouch" or "pouchdb" namespace for the docker work.  Maybe it's time to spin off the docker bits into a new micromodule with a name that actually reflects its purpose, like `gpii-couch-test-harness` or something shorter.  If we did that, we could pare down gpii-pouchdb to a smaller module that only supports the pouch-specific use cases we agree are valuable (standalone use in the browser, possibly node, depending on Cindy's input).  IMO this is a good discussion for the meeting tomorrow.

**Comment by the-t-in-rtf** [comment]
@cindyli, @amb26, @sgithens and I met today to discuss.  We agreed:

1.  That this pull should be withdrawn in favour of a pull against a new (empty) package that is only for the purpose of testing against the CouchDB REST API.
2. That (for environments like Windows 7), we should provide a workalike option where a Vagrant VM is provisioned with docker and the desired CouchDB container image, exposed on the same hostname and port.
3. That the port detection infrastructure should be replaced with manually specifying the port on container creation, and setting that to a single known port (`25984`, for example).

**Comment by the-t-in-rtf** [comment]
I have created [GPII-3530](https://issues.gpii.net/browse/GPII-3530) for the new repo, and [GPII-3531](https://issues.gpii.net/browse/GPII-3531) for the new pull.  I will close this one out and open a new pull shortly.

**Pull request closed without merging**